okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java

```diff
@@ -113,20 +116,30 @@ public static final class Builder {
     }
 
     /**
-     * No default. The POST URL for zipkin's <a href="https://zipkin.io/zipkin-api/#/">v2 api</a>,
+     * No default. See JavaDoc on {@link HttpEndpointSupplier} for implementation notes.
+     */
+    public Builder endpointSupplierFactory(HttpEndpointSupplier.Factory endpointSupplierFactory) {
+      if (endpointSupplierFactory == null) {
+        throw new NullPointerException("endpointSupplierFactory == null");
+      }
+      this.endpointSupplierFactory = endpointSupplierFactory;
+      return this;
+    }
+
+    /**
+     * No default. The POST HttpUrl for zipkin's <a href="https://zipkin.io/zipkin-api/#/">v2 api</a>,
      * usually "http://zipkinhost:9411/api/v2/spans"
      */
     // customizable so that users can re-map /api/v2/spans ex for browser-originated traces
     public Builder endpoint(String endpoint) {
       if (endpoint == null) throw new NullPointerException("endpoint == null");
-      HttpUrl parsed = HttpUrl.parse(endpoint);
-      if (parsed == null) throw new IllegalArgumentException("invalid POST url: " + endpoint);
-      return endpoint(parsed);
+      this.endpoint = endpoint;
```

Comment:
I think the `endpoint` is not needed anymore, is it? (we could use `endpointSupplierFactory` solely):

```java
this.endpointSupplierFactory = (HttpEndpointSupplier.Constant) endpoint -> endpoint;
```

Reply:
I should note where we are assigning this, that we need to retain the old endpoint or else we break the former (toBuilder()) functions. will make a comment!

Reply:
for example, if someone changes like .. toBuilder().endpointSupplierFactory(foo).build() we don't want to erase the (config) endpoint used to create the supplier

Reply:
added comments and edge case tests!

Reply:
Got it, thanks @codefromthecrypt !